drivers/flash/soc_flash_silabs_series2.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2018, Piotr Mienkowski
+ * Copyright (c) 2025 Silicon Laboratories Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT     silabs_series2_flash_controller
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/dma.h>
+#include <soc.h>
+#include <em_msc.h>
+
+#define LOG_LEVEL CONFIG_FLASH_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(flash_silabs);
+
+struct flash_silabs_data {
+	struct k_sem mutex;
+#if defined(CONFIG_SOC_FLASH_SILABS_S2_DMA_WRITE) || defined(CONFIG_SOC_FLASH_SILABS_S2_DMA_READ)
+	const struct device *dma_dev;
+	uint32_t dma_channel;
+#endif
+#if defined(CONFIG_SOC_FLASH_SILABS_S2_DMA_READ)
+	struct k_sem sync;
+	int sync_status;
+	struct dma_config dma_cfg;
+	struct dma_block_config dma_block_cfg;
+#endif
+};
+
+struct flash_silabs_config {
+	bool lpwrite;
+};
+
+static const struct flash_parameters flash_silabs_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+	.erase_value = 0xff,
+};
+
+static bool flash_silabs_read_range_is_valid(off_t offset, uint32_t size)
+{
+	return (offset + size) <= (CONFIG_FLASH_SIZE * 1024);
+}
+
+static bool flash_silabs_write_range_is_valid(off_t offset, uint32_t size)
+{
+	/* Note:
+	 * - A flash address to write to must be aligned to words.
+	 * - Number of bytes to write must be divisible by 4.
+	 */
+	return flash_silabs_read_range_is_valid(offset, size) && (offset % sizeof(uint32_t) == 0) &&
+	       (size % 4 == 0U);
```

Comment:
Is it possible to take advantage of the `IS_ALIGNED()` macro here? (defined in `include/zephyr/sys/util.h`)